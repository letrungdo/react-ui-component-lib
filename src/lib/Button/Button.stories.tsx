import { action } from "@storybook/addon-actions";
import { boolean, radios, text } from "@storybook/addon-knobs";
import React from "react";
import { Button } from ".";
import ImgUser from "../../../static/img_user.png";
import { FontType } from "../Label";

export default {
    title: "Core/Button",
};

export const Default = () => (
    <Button
        className="btn-link"
        onClick={action("on click")}
        fontFamily={radios("fontFamily", FontType, FontType.RobotoRegular)}
    >
        {text("label", "Button")}
    </Button>
);

export const ImageButton = () => (
    <Button
        disabled={boolean("disabled", false)}
        onClick={action("on click")}
        fontFamily={radios("fontFamily", FontType, FontType.RobotoRegular)}
    >
        <img style={{ width: 30 }} className="mr-3" src={ImgUser} />
        {text("label", "Button")}
    </Button>
);
