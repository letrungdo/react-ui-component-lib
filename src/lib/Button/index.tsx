import styled from "@emotion/styled";
import React from "react";
import { FontType } from "../Label";

interface Props {
    onClick?: (ev: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
    className?: string;
    fontFamily?: FontType;
    disabled?: boolean;
    btnRef?: string | ((instance: HTMLButtonElement | null) => void) | React.RefObject<HTMLButtonElement> | null;
    children?: React.ReactNode;
}

const Button = ({ onClick, className, disabled, btnRef, children, fontFamily }: Props) => (
    <ButtonStyle
        fontFamily={fontFamily}
        disabled={disabled}
        type="button"
        className={`btn ${className}`}
        onClick={onClick}
        ref={btnRef}
    >
        {children}
    </ButtonStyle>
);

Button.defaultProps = {
    className: "",
    fontFamily: FontType.RobotoRegular,
};

const ButtonStyle = styled.button`
    -webkit-tap-highlight-color: transparent;
    background-color: transparent;
    border: 0;
    border-radius: 0;
    cursor: pointer;
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 99;
    font-family: ${(props: Props) => props.fontFamily};
    &:active {
        opacity: 0.7;
    }
    &:disabled {
        cursor: unset;
        &:active {
            opacity: 1;
        }
    }
`;

export { Button };
