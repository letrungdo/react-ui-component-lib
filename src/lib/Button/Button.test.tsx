import React from "react";
import { Button } from "../Button";
import { mount, shallow } from "enzyme";

describe("Button", () => {
    const TEXT = "this is text";
    const ev = jest.fn();

    beforeEach(() => {
        ev.mockReset();
    });

    test("Has text", () => {
        const elm = mount(<Button>{TEXT}</Button>);

        expect(elm).toMatchSnapshot();
    });

    test("On Click", () => {
        const elm = shallow(<Button onClick={ev}>Click me</Button>);

        elm.simulate("click");
        expect(ev.mock.calls.length).toBe(1);

        expect(elm).toMatchSnapshot();
    });

    test("Disabled", () => {
        const elm = shallow(<Button disabled={true}>Disabled</Button>);

        expect(elm.is("[disabled]")).toBe(true);

        expect(elm).toMatchSnapshot();
    });
});
