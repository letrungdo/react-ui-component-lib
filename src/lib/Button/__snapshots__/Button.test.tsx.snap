// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Disabled 1`] = `
<Styled(button)
  className="btn "
  disabled={true}
  fontFamily="RobotoRegular"
  type="button"
>
  Disabled
</Styled(button)>
`;

exports[`Button Has text 1`] = `
.emotion-0 {
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  border: 0;
  border-radius: 0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 99;
  font-family: RobotoRegular;
}

.emotion-0:active {
  opacity: 0.7;
}

.emotion-0:disabled {
  cursor: unset;
}

.emotion-0:disabled:active {
  opacity: 1;
}

<Button
  fontFamily="RobotoRegular"
>
  <Styled(button)
    className="btn "
    fontFamily="RobotoRegular"
    type="button"
  >
    <button
      className="btn  emotion-0"
      fontFamily="RobotoRegular"
      type="button"
    >
      this is text
    </button>
  </Styled(button)>
</Button>
`;

exports[`Button On Click 1`] = `
<Styled(button)
  className="btn "
  fontFamily="RobotoRegular"
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  type="button"
>
  Click me
</Styled(button)>
`;
